Release/4.5.2

abbydomi opened the pull request:
This pull request updates the Applivery SDK version and improves the logic for checking app updates, particularly around build numbers and versioning. It also clarifies logging messages and corrects test data to better reflect real-world scenarios.

**Versioning and Metadata Updates:**
- Updated the SDK version to `4.5.2` in both the `Applivery.podspec` and `AppliverySDK.swift` files to ensure consistency across the project. [[1]](diffhunk://#diff-d91b8c9d56367343821559d239b096f070296f86f6038e5200abc6314df8b460L3-R3) [[2]](diffhunk://#diff-f86d3e567fed0c23b2547250bbe6181e247a7794549efc32abd602e79511da51L121-R121)

**Update Logic Improvements:**
- Changed the logic in `UpdateService` so that OTA updates are now checked using the app's build number instead of the app version, aligning with typical OTA update practices.
- Updated log messages in `UpdateService` to clarify whether checks are comparing app version or build number, making debugging and maintenance easier. [[1]](diffhunk://#diff-142054b531cf72e61e7607ec5f78e21e2bbd59e6f2c79ce37717c0364b56cc38L150-R150) [[2]](diffhunk://#diff-142054b531cf72e61e7607ec5f78e21e2bbd59e6f2c79ce37717c0364b56cc38L169-R169)

**Test Adjustments:**
- Adjusted test data in `UpdateServiceTests.swift` to use a realistic build number (`101`) for `lastBuildVersion`, and removed a redundant version stub, ensuring tests accurately reflect the new update logic.

abbydomi committed:
feat: add disk space error handling and build size tracking

Introduced new UpdateError cases for disk space issues and updated SDKData and ConfigPersister to store and retrieve the last build size. These changes improve update reliability by allowing the SDK to detect and handle insufficient disk space scenarios.

abbydomi committed:
feat: Add disk space check before download in LoginService

Introduced a check for available device storage before initiating a download in LoginService. Refactored SessionPersister to use a protocol, allowing for dependency injection and improved testability.

abbydomi committed:
feat: Add mocks and tests for LoginService and update config mocks

Added new mock classes (DownloadServiceMock, LoginRepositoryMock, MockSessionPersister) and comprehensive LoginServiceTests to improve test coverage. Updated ConfigMockData, ConfigServiceMock, and UpdateServiceTests to include the lastBuildSize property for more accurate configuration mocking.

abbydomi committed:
feat: Move deviceAvailableSpace to App and update usage

Refactored deviceAvailableSpace from LoginService to App, making it part of the AppProtocol. Updated LoginService to use app.deviceAvailableSpace() and removed the local implementation. This centralizes device storage checks and improves code organization.

abbydomi committed:
feat: Add deviceAvailableSpace stub to AppMock

Introduced a stub property and method for deviceAvailableSpace in AppMock to allow tests to simulate available device storage. This supports testing features that depend on device storage availability.

abbydomi committed:
feat: Improve assertions and test setup in LoginServiceTests

Refines test assertions to check result type and error explicitly instead of matching the entire result. Updates test setup by stubbing device available space and open URL result where necessary to ensure accurate test conditions.

abbydomi committed:
style: remove unnecessary blank lines in App.swift

Cleaned up extra blank lines in the App class implementation to improve code readability and maintain consistency.

abbydomi committed:
style: remove commented-out save method in SessionPersister

Cleaned up SessionPersister by deleting an unused, commented-out save method to improve code readability.

abbydomi committed:
refactor: SDKData mock usage in tests

Added a static mock property to SDKData and updated LoginServiceTests to use SDKData.mock instead of duplicating mock initialization. This improves test maintainability and reduces code duplication.

abbydomi committed:
refactor: Show error alerts on main thread in LoginService

Wrapped showErrorAlert calls in DispatchQueue.main.async to ensure UI updates occur on the main thread when handling storage errors during login.

abbydomi committed:
Fix: OTA update check to use build number instead of version

Updated the OTA update logic in UpdateService to compare the app's build number with the last build version, ensuring correct update prompts. Adjusted related test data and improved log messages for clarity.

abbydomi committed:
build: Bump SDK version to 4.5.2

Updated the podspec and internal sdkVersion to 4.5.2 to reflect the new release.

woa1mu commented on AppliverySDK/Services/UpdateService.swift:
```diff
@@ -197,14 +197,16 @@ final class UpdateService: UpdateServiceProtocol {
         }
 
-        if checkOtaUpdate(updateConfig.config, version: appVersion) {
+        if checkOtaUpdate(updateConfig.config, version: appBuildNumber) {
```
I propose to change the signature of `checkOtaUpdate` and rename parameter **version** to **buildNumber**. 
This ensures clarity and avoids future confusion.

abbydomi replied:
I agree, this is a good change for now

woa1mu commented on AppliverySDK/Services/UpdateService.swift:
```diff
@@ -166,7 +166,7 @@ final class UpdateService: UpdateServiceProtocol {
         }
 
-        logInfo("[checkOtaUpdate] - Checking if app version: \(version) is older than last build version: \(lastVersion)")
+        logInfo("[checkOtaUpdate] - Checking if build number: \(version) is older than last build version: \(lastVersion)")
```
Maybe in general streamline the usage of **buildNumber** and **version**.

abbydomi committed:
Refactor: Rename 'version' to 'buildNumber' in checkOtaUpdate

Refactored the checkOtaUpdate method and its usages to use 'buildNumber' instead of 'version' for clarity and consistency with the parameter's purpose.

abbydomi committed:
refactor: Rename parameter in checkOtaUpdate mock methods

Changed the parameter name from 'version' to 'buildNumber' in the checkOtaUpdate methods of MockUpdateService and UpdateServiceMock to improve clarity and consistency with expected usage.

woa1mu commented on AppliverySDK/Services/UpdateService.swift:
```diff
@@ -166,8 +166,8 @@ final class UpdateService: UpdateServiceProtocol {
-        logInfo("[checkOtaUpdate] - Checking if app version: \(version) is older than last build version: \(lastVersion)")
-        if self.isOlder(version, minVersion: lastVersion) {
+        logInfo("[checkOtaUpdate] - Checking if build number: \(buildNumber) is older than last build version: \(lastVersion)")
+        if self.isOlder(buildNumber, minVersion: lastVersion) {
```
@abbydomi might this be a problem now?
comparing "101" with "2.0.1" (?) or is *lastVersion* actully *lastBuildNumber*?

(https://github.com/applivery/applivery-ios-sdk/pull/100#discussion_r2434767579)

woa1mu replied:
@abbydomi can we merge it like this? I'd like to integrate your fix and test it

abbydomi committed:
fix: Change update checks to compare version and build number

Updated UpdateService and related mocks to compare both semantic version and build number when checking for updates. The isUpToDate, checkForceUpdate, and checkOtaUpdate methods now use a new helper to ensure more accurate update detection, and their signatures have been updated accordingly.

abbydomi committed:
refactor: Update checking using build number comparison

Simplifies version checking and removes the previously added but not properly checked semantic app version

abbydomi committed:
refactor: update mock methods to use buildNumber parameter

Renamed the 'version' parameter to 'buildNumber' in checkForceUpdate and checkOtaUpdate methods in MockUpdateService and UpdateServiceMock to match updated method signatures.

abbydomi committed:
feat: comprehensive tests for UpdateService checkUpdate

Expanded UpdateServiceTests with new test cases covering forced updates, forced updates not needed, OTA updates, and OTA updates not needed. This improves coverage and ensures correct behavior for various update scenarios.

abbydomi committed:
chore: clarify build number usage in update methods

Updated documentation for isUpToDate() and checkForUpdates() to specify that only the build number is considered when determining if the app is up to date.

woa1mu commented:
@FranAlarza can we have this released? 🙏

abbydomi committed:
Merge branch 'develop' into feature/apple-51-improve-error-feedback-when-app-install-fails-due-to

EduardoParadaApplivery committed:
Merge pull request #88 from applivery/feature/apple-51-improve-error-feedback-when-app-install-fails-due-to

Feature/apple 51 improve error feedback when app install fails due to

abbydomi committed:
Merge branch 'develop' into release/4.5.2

abbydomi committed:
refactor: remove unused variable

Eliminated the unused 'version' variable assignment in the handleAppWillEnterForeground method to clean up the code.

abbydomi committed:
Add lastBuildSize to UpdateService test cases

Included the lastBuildSize property with a nil value in relevant UpdateServiceTests cases to reflect updated model requirements.

abbydomi committed:
chore: add provisional logs for tests

abbydomi committed:
chore: Increase waitUntil timeout in UpdateServiceTests

The default timeout for the waitUntil helper was increased from 2.0 to 10.0 seconds to allow more time for conditions to be met during tests, reducing potential flakiness.

abbydomi committed:
chore: Increase waitUntil timeout in UpdateServiceTests

Raised the default timeout for the waitUntil helper from 10 to 50 seconds to allow more time for conditions to be met during tests.

abbydomi committed:
chore: Remove provisional logs

abbydomi committed:
refactor: force update test to use OTA alert spy

Renamed variables from forceUpdateConfig to otaConfig and updated test logic to check appMock.spyOtaAlert.called instead of appMock.spyForceUpdateCalled. Also reduced the default timeout in waitUntil from 50.0 to 20.0 seconds for faster test execution.
